kotlin_engine/src/main/java/com/github/paylike/kotlin_engine/viewmodel/PaylikeEngine.kt

```diff
@@ -0,0 +1,287 @@
+package com.github.paylike.kotlin_engine.viewmodel
+
+import android.util.Log
+import com.github.paylike.kotlin_client.PaylikeClient
+import com.github.paylike.kotlin_client.domain.dto.payment.request.PaymentData
+import com.github.paylike.kotlin_client.domain.dto.payment.request.card.ExpiryDto
+import com.github.paylike.kotlin_client.domain.dto.payment.request.card.PaylikeCardDto
+import com.github.paylike.kotlin_client.domain.dto.payment.request.integration.PaymentIntegrationDto
+import com.github.paylike.kotlin_client.domain.dto.payment.request.test.PaymentTestDto
+import com.github.paylike.kotlin_client.domain.dto.payment.response.PaylikeClientResponse
+import com.github.paylike.kotlin_client.domain.dto.tokenize.request.TokenizeData
+import com.github.paylike.kotlin_client.domain.dto.tokenize.request.TokenizeTypes
+import com.github.paylike.kotlin_engine.helper.PaylikeEngineError
+import com.github.paylike.kotlin_engine.helper.exceptions.*
+import com.github.paylike.kotlin_engine.model.repository.EngineRepository
+import com.github.paylike.kotlin_engine.model.service.ApiMode
+import com.github.paylike.kotlin_luhn.PaylikeLuhn
+import com.github.paylike.kotlin_money.PaymentAmount
+import com.github.paylike.kotlin_request.exceptions.PaylikeException
+import java.util.*
+import java.util.function.Consumer
+import kotlin.reflect.full.superclasses
+import kotlinx.coroutines.async
+import kotlinx.coroutines.coroutineScope
+
+/** Paylike engine */
+// TODO KDoc of engine
+class PaylikeEngine(private val merchantId: String, private val apiMode: ApiMode) : Observable() {
+
+    val repository: EngineRepository = EngineRepository()
+
+    var currentState: EngineState = EngineState.WAITING_FOR_INPUT
+        private set
+
+    var error: PaylikeEngineError? = null
+        private set
+
+    private val apiService: PaylikeClient = PaylikeClient()
+
+    private val log: Consumer<Any> = Consumer { Log.e("Engine logger", it.toString()) }
+
+    /**
+     * Execute api calls and create the necessary data for the [EngineRepository.paymentRepository]
```

Comment:
I am not sure if this function should work like this. Tokenization of the card information is of course required, but how will the user edit the paymentData?

Users should not be able to set anytinhg in the repository directly. So for example, you wanna add an amount, how will you do that?

Reply:
The startPayment function adds the amount. Then a renaming can help this misunderstanding.

Reply:
I refactor this functionality to be more modular